genic/power.c

```diff
@@ -72,113 +116,93 @@ void read_power_table(void)
-
-        if(!(fd = fopen(buf, "r")))
-        {
-            endrun(1, "can't read input spectrum in file '%s' on task %d\n", buf, ThisTask);
-        }
-
-        i = 0;
         do
         {
-            double k, p;
             char buffer[1024];
-            char * retval = fgets(buffer, 1024, fd);
+            char * line = fgets(buffer, 1024, fd);
             /*Happens on end of file*/
-            if(!retval)
+            if(!line)
                 break;
-            retval = strtok(buffer, " \t");
+            char * retval = strtok(line, " \t");
             if(!retval || retval[0] == '#')
                 continue;
-            k = atof(retval);
-            if(k < 0 && !InputInLog10) {
-                message(1, "some input k is negative, guessing the file is in log10 units\n");
-                InputInLog10 = 1;
-            }
-            retval = strtok(NULL, " \t");
-            if(!retval)
-                endrun(1,"Incomplete line in power spectrum: %s\n",buffer);
-            p = atof(retval);
-            logk_tab[i] = k;
-            if (!InputInLog10) {
-                k = log10(k);
-                p = log10(p);
-            }
-
-            k -= log10(InputSpectrum_UnitLength_in_cm / UnitLength_in_cm);	/* convert to h/Kpc */
-            logk_tab[i] = k;
-
-            p += 3 * log10(InputSpectrum_UnitLength_in_cm / UnitLength_in_cm);	/* convert to Kpc/h  */
-            logD_tab[i] = p;
+            double k = atof(retval);
+            parse_line(i, k, line, out_tab, &InputInLog10, scale);
             i++;
         }
         while(1);
 
         fclose(fd);
-
     }
 
-    MPI_Bcast(logk_tab, 2*NPowerTable, MPI_DOUBLE, 0, MPI_COMM_WORLD);
-    mat_intp = gsl_interp_alloc(gsl_interp_cspline,NPowerTable);
-    mat_intp_acc = gsl_interp_accel_alloc();
-    gsl_interp_init(mat_intp,logk_tab, logD_tab,NPowerTable);
+    MPI_Bcast(out_tab->logk, (ncols+1)*out_tab->Nentry, MPI_DOUBLE, 0, MPI_COMM_WORLD);
+    for(j=0; j<ncols; j++) {
+        out_tab->mat_intp[j] = gsl_interp_alloc(gsl_interp_cspline,out_tab->Nentry);
+        out_tab->mat_intp_acc[j] = gsl_interp_accel_alloc();
+        gsl_interp_init(out_tab->mat_intp[j],out_tab->logk, out_tab->logD[j],out_tab->Nentry);
+    }
 }
 
-void initialize_powerspectrum(void)
+int initialize_powerspectrum(int ThisTask, double InitTime, double UnitLength_in_cm_in, Cosmology * CPin, struct power_params * ppar)
 {
-    if(WhichSpectrum == 2)
-        read_power_table();
+    if(ppar->WhichSpectrum == 2) {
+        read_power_table(ThisTask, ppar->FileWithInputSpectrum, 1, &power_table, ppar->SpectrumLengthScale, parse_power);
+        transfer_table.Nentry = 0;
+        if(ppar->DifferentTransferFunctions)
+            read_power_table(ThisTask, ppar->FileWithTransferFunction, MAXCOLS, &transfer_table, ppar->SpectrumLengthScale, parse_transfer);
+        message(0, "Power spectrum rows: %d, Transfer: %d\n", power_table.Nentry, transfer_table.Nentry);
```

Comment:
Is it more convenient  to have one multi column file or is it more convinient to have several files?

Reply:
CAMB produces one multi-column file. Or do you mean the internal structure in the code?
One structure is more convenient because the transfer functions have the same k (but the matter power does not) and are normalised the same.

Reply:
I see. the idea is to read the output of CAMB directly.